src/installed_dune_file.ml

```diff
@@ -21,33 +21,54 @@ let parse_sub_systems sexps =
   |> Sub_system_name.Map.mapi ~f:(fun name (_, version, data) ->
     let (module M) = Jbuild.Sub_system_info.get name in
     Syntax.check_supported M.syntax version;
-    let parsing_context =
-      Univ_map.singleton (Syntax.key M.syntax)
-        (* This is wrong, see #909 *)
-        (0, 0)
-    in
     M.T (Sexp.Of_sexp.parse M.parse parsing_context data))
 
 let of_sexp =
   let open Sexp.Of_sexp in
   let version =
     plain_string (fun ~loc -> function
-      | "1" -> ()
-      | _ ->
+      | "1" | "2" -> ()
+      | v ->
         of_sexp_errorf loc
-          "Unsupported version, only version 1 is supported")
+          "Unsupported version %S, only version 1 is supported" v)
   in
   sum
     [ "dune",
       (version >>= fun () ->
-       list raw >>| fun l ->
-       parse_sub_systems l)
+       get_all >>= fun parsing_context ->
+       list raw >>|
+       parse_sub_systems ~parsing_context)
     ]
 
 let load fname =
-  Sexp.Of_sexp.parse of_sexp Univ_map.empty (Io.Sexp.load ~mode:Single fname)
+  Io.with_lexbuf_from_file fname ~f:(fun lexbuf ->
+    let (version_loc, version) =
+      let bad_dune_file = "Unable to read (dune x.y ..) line file" in
+      let rec loop = function
+        | [_; _; _] as a -> List.rev a
+        | acc ->
+          begin match (Sexp.Lexer.token lexbuf : Sexp.Lexer.Token.t) with
+          | Eof ->
+            Loc.fail (Loc.in_file (Path.to_string fname)) "%s" bad_dune_file
```

Comment:
You can return `List.rev acc`, the error will be caught by the `match loop [] with ...`.